jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java

```diff
@@ -55,14 +57,29 @@ public HostHeaderCustomizer(String serverName)
      */
     public HostHeaderCustomizer(String serverName, int serverPort)
     {
-        this.serverName = Objects.requireNonNull(serverName);
-        this.serverPort = serverPort;
+        String host = Objects.requireNonNull(serverName);
+        if (serverPort > 0)
+            host += ":" + serverPort;
+        hostValue = host;
     }
 
     @Override
     public void customize(Connector connector, HttpConfiguration channelConfig, Request request)
     {
-        if (request.getHeader("Host") == null)
-            request.setAuthority(serverName, serverPort);  // TODO set the field as well?
+        String hostHeaderValue = request.getHeader("Host");
+
+        // No Host Header
+        if (request.getHttpVersion().getVersion() != HttpVersion.HTTP_1_1.getVersion() &&
+            hostHeaderValue == null)
+        {
+            if (request.getHttpURI().isAbsolute())
+            {
+                request.getHttpFields().put(HttpHeader.HOST, request.getHttpURI().getAuthority());
+            }
+            else
+            {
+                request.getHttpFields().put(HttpHeader.HOST, hostValue);
+            }
+        }
```

Comment:
Since getHeader is a linear lookup, it is best to avoid doing it if possible, and then to use efficient non string lookup:
```suggestion
        // If not 1.1 and no Host Header
        if (request.getHttpVersion().getVersion() != HttpVersion.HTTP_1_1.getVersion() &&
            !request.getHttpFields().contains(HttpHeader.HOST))
        {
            if (request.getHttpURI().isAbsolute())
            {
                request.getHttpFields().put(HttpHeader.HOST, request.getHttpURI().getAuthority());
            }
            else
            {
                request.getHttpFields().put(HttpHeader.HOST, hostValue);
            }
        }
```